ntp-proto/src/packet/crypto.rs

```diff
@@ -0,0 +1,112 @@
+use aes_siv::{siv::Aes128Siv, Key, KeyInit};
+use rand::Rng;
+use tracing::error;
+
+use super::{extensionfields::ExtensionField, PacketParsingError};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EncryptionResult {
+    pub nonce: Vec<u8>,
+    pub ciphertext: Vec<u8>,
+}
+
+pub trait Cipher: Sync + Send {
+    fn encrypt(
+        &self,
+        plaintext: &[u8],
+        associated_data: &[u8],
+    ) -> std::io::Result<EncryptionResult>;
+    fn decrypt(
+        &self,
+        nonce: &[u8],
+        ciphertext: &[u8],
+        associated_data: &[u8],
+    ) -> Result<Vec<u8>, PacketParsingError>;
+}
+
+pub trait CipherProvider {
+    fn get(&self, context: &[ExtensionField<'_>]) -> Option<&dyn Cipher>;
+}
+
+impl CipherProvider for () {
+    fn get<'a>(&self, _context: &[ExtensionField<'_>]) -> Option<&dyn Cipher> {
+        None
+    }
+}
```

Comment:
this looks weird at the use site. Can't we have something with a better name? e.g. `struct NoCipher;` I think would look much better at the use site

```diff
- let rec_packet = NtpPacket::deserialize(&buf, &()).unwrap();
+ let rec_packet = NtpPacket::deserialize(&buf, &NoCipher).unwrap();
```

Reply:
Done